src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,421 @@
+        };
+    linker
+        .func_wrap(module_name, "font_get_glyph", font_get_glyph)
+        .ok()?;
+
+    // fn font_get_scale(font: u32, x_scale: *mut i32, y_scale: *mut i32);
+    // Returns the scale of the current font.
+    // Just return the upem as rustybuzz has no scale.
+    let font_get_scale =
+        |mut caller: Caller<'_, ShapingData>, _: u32, x_scale: u32, y_scale: u32| {
+            let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+            let upem = caller.data().font.units_per_em();
+
+            memory.write(
+                &mut caller.as_context_mut(),
+                x_scale as usize,
+                &upem.to_le_bytes(), // le or be?
+            );
+            memory.write(
+                &mut caller.as_context_mut(),
+                y_scale as usize,
+                &upem.to_le_bytes(),
+            );
+        };
+    linker
+        .func_wrap(module_name, "font_get_scale", font_get_scale)
+        .ok()?;
+
+    // fn font_get_glyph_extents(font: u32, glyph: u32, extents: *mut CGlyphExtents) -> bool;
+    // Returns the glyph's extents for the given glyph ID at current scale and variation settings.
+    let font_get_glyph_extents =
+        |mut caller: Caller<'_, ShapingData>, _: u32, glyph: u32, extents: u32| -> u32 {
+            let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+            let mut glyph_extents = hb_glyph_extents_t::default();
+            let ret = caller
+                .data()
+                .font
+                .glyph_extents(GlyphId(glyph as u16), &mut glyph_extents);
+            if ret {
+                // WASM is little endian. MacOS is little endian, I think.
+                // Maybe should use bytemuck as a dependency instead.
+                let glyph_extents = unsafe { std::mem::transmute(glyph_extents) };
+                memory.write(caller.as_context_mut(), extents as usize, glyph_extents);
+            }
+
+            ret as u32
+        };
+    linker
+        .func_wrap(
+            module_name,
+            "font_get_glyph_extents",
+            font_get_glyph_extents,
+        )
+        .ok()?;
+
+    // fn font_glyph_to_string(font: u32, glyph: u32, str: *const u8, len: u32);
+    // Copies the name of the given glyph, or, if no name is available, a string of the form gXXXX into the given string.
+    let font_glyph_to_string =
+        |mut caller: Caller<'_, ShapingData>, _: u32, glyph: u32, str: u32, _len: u32| {
+            // what am I supposed to do with `len` here? Should I trim the name to `len` length somehow?
+            let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+            let name = caller
+                .data()
+                .font
+                .glyph_name(GlyphId(glyph as u16))
+                .map(ToOwned::to_owned)
+                .unwrap_or(format!("g{glyph:4}"));
+
+            let name = CString::new(name).unwrap();
```

Comment:
Are you sure it is C string and not just `[i8; N]`?

Reply:
`harfbuzz-wasm` assumes it is a `Cstring` : https://github.com/harfbuzz/harfbuzz/blob/e079dd20304bacd8d753eee2ceacbaa1a3978a34/src/wasm/rust/harfbuzz-wasm/src/lib.rs#L459

Reply:
Good. Just wanted to make sure. Then we can skip length I guess.